File: leetcode/Hard/csharp/2872. Maximum Number of K-Divisible Components.cs
/**
 * https://leetcode.com/problems/maximum-number-of-k-divisible-components/description/?envType=daily-question&envId=2025-11-28
 * Автор: Дуплей Максим Игоревич - AGLA
 * ORCID: https://orcid.org/0009-0007-7605-539X
 * GitHub: https://github.com/QuadDarv1ne/
 * 
 * Решение задачи "Maximum Number of K-Divisible Components" на C#
 * 
 * Полезные ссылки:
 * 1. Telegram ❃ Хижина программиста Æ: https://t.me/hut_programmer_07
 * 2. Telegram №1 @quadd4rv1n7
 * 3. Telegram №2 @dupley_maxim_1999
 * 4. Rutube канал: https://rutube.ru/channel/4218729/
 * 5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
 * 6. YouTube канал: https://www.youtube.com/@it-coders
 * 7. ВК группа: https://vk.com/science_geeks
 */

using System;
using System.Collections.Generic;

public class Solution {
    private List<int>[] graph;
    private int[] values;
    private int k;
    private int components;
    
    public int MaxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        // Инициализация графа
        this.graph = new List<int>[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new List<int>();
        }
        foreach (var edge in edges) {
            int u = edge[0], v = edge[1];
            graph[u].Add(v);
            graph[v].Add(u);
        }
        
        this.values = values;
        this.k = k;
        this.components = 0;
        
        DFS(0, -1);
        return components;
    }
    
    private long DFS(int node, int parent) {
        long total = values[node];
        
        foreach (int neighbor in graph[node]) {
            if (neighbor != parent) {
                total += DFS(neighbor, node);
            }
        }
        
        if (total % k == 0) {
            components++;
        }
        
        return total;
    }
}

File: leetcode/Hard/python/2872. Maximum Number of K-Divisible Components.py
'''
https://leetcode.com/problems/maximum-number-of-k-divisible-components/description/?envType=daily-question&envId=2025-11-28
Автор: Дуплей Максим Игоревич - AGLA
ORCID: https://orcid.org/0009-0007-7605-539X
GitHub: https://github.com/QuadDarv1ne/

Решение задачи "Maximum Number of K-Divisible Components"

Полезные ссылки:
1. Telegram ❃ Хижина программиста Æ: https://t.me/hut_programmer_07
2. Telegram №1 @quadd4rv1n7
3. Telegram №2 @dupley_maxim_1999
4. Rutube канал: https://rutube.ru/channel/4218729/
5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
6. YouTube канал: https://www.youtube.com/@it-coders
7. ВК группа: https://vk.com/science_geeks
'''

class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        # Строим граф
        graph = [[] for _ in range(n)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        self.components = 0
        
        def dfs(node: int, parent: int) -> int:
            # Начинаем с значения текущего узла
            total = values[node]
            
            # Обрабатываем всех соседей кроме родителя
            for neighbor in graph[node]:
                if neighbor != parent:
                    total += dfs(neighbor, node)
            
            # Если сумма делится на k, увеличиваем счетчик компонент
            if total % k == 0:
                self.components += 1
            
            return total
        
        # Запускаем DFS из корня (узла 0)
        dfs(0, -1)
        return self.components